pkg/e2e/infra/infra.go

```diff
@@ -151,6 +153,98 @@ var _ = g.Describe("[Feature:Machines] Managed cluster should", func() {
+
+		machineSets, err := getMachineSets(client)
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(len(machineSets)).To(o.BeNumerically(">", 2))
+		machineSet := machineSets[0]
+		initialReplicasMachineSet := int(pointer.Int32PtrDerefOr(machineSet.Spec.Replicas, 0))
+		scaleOut := 3
+		scaleIn := initialReplicasMachineSet
+		originalReplicas := initialReplicasMachineSet
+		clusterGrowth := scaleOut - originalReplicas
+		clusterDecrease := scaleOut - scaleIn
+		intermediateClusterSize := initialClusterSize + clusterGrowth
+		finalClusterSize := initialClusterSize + clusterGrowth - clusterDecrease
+
+		g.By(fmt.Sprintf("scaling out a %q machineSet to %d replicas", machineSet, scaleOut))
+		err = scaleMachineSet(machineSet.Name, scaleOut)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By(fmt.Sprintf("waiting for cluster to grow %d nodes. Size should be %d", clusterGrowth, intermediateClusterSize))
+		waitForClusterSizeToBeHealthy(intermediateClusterSize)
+
+		g.By(fmt.Sprintf("scaling in %q machineSet to %d replicas", machineSet, scaleIn))
+		err = scaleMachineSet(machineSet.Name, scaleIn)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By(fmt.Sprintf("waiting for cluster to decrease %d nodes. Final size should be %d nodes", clusterDecrease, finalClusterSize))
+		waitForClusterSizeToBeHealthy(finalClusterSize)
+	})
+
+	g.It("grow and decrease when scaling different machineSets simultaneously", func() {
+		client, err := e2e.LoadClient()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		scaleOut := 3
+
+		g.By("checking initial cluster size")
+		initialClusterSize, err := getClusterSize(client)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("getting worker machineSets")
+		machineSets, err := getMachineSets(client)
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(len(machineSets)).To(o.BeNumerically(">", 2))
+		machineSet0 := machineSets[0]
+		initialReplicasMachineSet0 := int(pointer.Int32PtrDerefOr(machineSet0.Spec.Replicas, 0))
+		machineSet1 := machineSets[1]
+		initialReplicasMachineSet1 := int(pointer.Int32PtrDerefOr(machineSet1.Spec.Replicas, 0))
+
+		g.By(fmt.Sprintf("scaling %q from %d to %d replicas", machineSet0.Name, initialReplicasMachineSet0, scaleOut))
+		err = scaleMachineSet(machineSet0.Name, scaleOut)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By(fmt.Sprintf("scaling %q from %d to %d replicas", machineSet1.Name, initialReplicasMachineSet1, scaleOut))
+		err = scaleMachineSet(machineSet1.Name, scaleOut)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		o.Eventually(func() bool {
+			nodes, err := getNodesFromMachineSet(client, machineSet0)
+			o.Expect(err).NotTo(o.HaveOccurred())
+			if len(nodes) != scaleOut {
+				return false
+			}
+			return nodesAreReady(nodes)
+		}, e2e.WaitLong, 5*time.Second).Should(o.BeTrue())
+
+		o.Eventually(func() bool {
+			nodes, err := getNodesFromMachineSet(client, machineSet1)
+			o.Expect(err).NotTo(o.HaveOccurred())
+			if len(nodes) != scaleOut {
```

Comment:
I think this can be reduced to: `return len(nodes) == scaleOut && nodesAreReady(nodes)`.